main_test.go

```diff
@@ -2,59 +2,241 @@ package main
+	backend := gcp.NewBackend(gcpSettings, hedwigLogger)
+	msgList := []hedwig.MessageTypeMajorVersion{{
+		MessageType:  "user-created",
+		MajorVersion: 1,
+	}}
+	var s3 ProcessSettings
+	var s2 gcp.Settings
+	storageBackend := firehoseGcp.NewBackend(&storage.Client{})
+	encoderDecoder := firehoseProtobuf.FirehoseEncoderDecoder{}
+	f, err := NewFirehose(backend, &encoderDecoder, msgList, storageBackend, s2, s3, hedwigLogger)
+	assert.Equal(s.T(), nil, err)
+	assert.NotNil(s.T(), f)
 }
 
-func (b *FirehoseBackendMock) ReadFile(ctx context.Context, readBucket string, readLocation string) ([]byte, error) {
-	return []byte("abc"), nil
+func (s *GcpTestSuite) TestFirehoseFollowerIntegration() {
+	var hedwigLogger hedwig.Logger
+	backend := gcp.NewBackend(s.pubSubSettings, hedwigLogger)
+	// maybe just user-created?
+	msgList := []hedwig.MessageTypeMajorVersion{{
+		MessageType:  "user-created",
+		MajorVersion: 1,
+	}}
+	s3 := ProcessSettings{
+		MetadataBucket: "some-metadata-bucket",
+		StagingBucket:  "some-staging-bucket",
+		OutputBucket:   "some-output-bucket",
+		FlushAfter:     2,
+	}
+	var s2 gcp.Settings
+	storageBackend := firehoseGcp.NewBackend(s.storageClient)
+	msgTypeUrls := map[hedwig.MessageTypeMajorVersion]string{
+		{MessageType: "user-created", MajorVersion: 1}: "type.googleapis.com/standardbase.hedwig.UserCreatedV1",
+	}
+	encoderDecoder := firehoseProtobuf.NewFirehoseEncodeDecoder(msgTypeUrls)
+	f, err := NewFirehose(backend, encoderDecoder, msgList, storageBackend, s2, s3, hedwigLogger)
+	s.Require().NoError(err)
+
+	routing := map[hedwig.MessageTypeMajorVersion]string{
+		{
+			MessageType:  "user-created",
+			MajorVersion: 1,
+		}: "dev-user-created-v1",
+	}
+	pubEncoderDecoder, err := protobuf.NewMessageEncoderDecoder(
+		[]proto.Message{&UserCreatedV1{}},
+	)
+	s.Require().NoError(err)
+	publisher := hedwig.NewPublisher(backend, pubEncoderDecoder, routing)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+
+	defer cancel()
+
+	userId := "C_1234567890123456"
+	data := UserCreatedV1{UserId: &userId}
+	message, err := hedwig.NewMessage("user-created", "1.0", map[string]string{"foo": "bar"}, &data, "myapp")
+	s.Require().NoError(err)
+	_, err = publisher.Publish(ctx, message)
+	s.Require().NoError(err)
+
+	_, err = publisher.Publish(ctx, message)
+	s.Require().NoError(err)
+
+	go f.RunFollower(ctx)
+
+	// stop test after 5sec, should finish by then
+	timerCh := time.After(5 * time.Second)
+outer:
+	for {
+		select {
+		case <-timerCh:
+			it := s.storageClient.Bucket("some-staging-bucket").Objects(context.Background(), nil)
+			userCreatedObjs := []string{}
+			for {
+				attrs, err := it.Next()
+				if err == iterator.Done {
+					break
+				}
+				s.Require().NoError(err)
+				// check that file under message folder
+				if strings.Contains(attrs.Name, "user-created/1") {
+					userCreatedObjs = append(userCreatedObjs, attrs.Name)
+				}
+			}
+			assert.Equal(s.T(), 2, len(userCreatedObjs))
```

Comment:
this is good, but should we perhaps also try to _read_ the contents of the file to make sure the message published was actually what is written to the disk?

Reply:
Or is that a follow up PR?

Reply:
yeah I can add to the test